### ISSUE TITLE
Optionally derive `IntoBytes` for `repr(C)` generic types using `Unaligned` constraints on fields 

### ISSUE THREAD

**Issue by reinerp** [google/zerocopy#825]
The padding check on fields can be skipped if all fields are known to implement Unaligned.

**Comment by joshlf** [comment]
In general I agree that this is viable. Can you give an example of a type which we currently reject but which you think should be accepted? Zerocopy is available in the Rust playground, so you can experiment with it in the playground. I.e.:

```rust
#[derive(AsBytes)]
#[repr(C)]
struct Foo {
    ...
}
```

**Comment by reinerp** [comment]
Ah sorry, I might have been too brief (was typing on my phone). The word "generic" is key: deriving of AsBytes currently rejects all generic types, but with my proposed approach it wouldn't have to.

Here's an example: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=2fa687834d9efd0c52f3ba629271096a

```
use zerocopy::*;

#[derive(AsBytes)]
#[repr(C)]
struct S<T: Unaligned> {
    x: u8,
    y: T,
}
```

**Comment by joshlf** [comment]
Ah I see. Yeah this seems viable! Do you have a specific use case for this, or is it just a curiosity? It's definitely something we'd want to do at some point, but if there is a specific use case, that could affect our prioritization.

**Comment by reinerp** [comment]
I do, but I unfortunately don't have a public repo to share yet.

I'm working on an Rust codegen from an IDL (similar to CapNProto). For the IDL I decided to make all types unaligned, since alignment for wire formats hurts (wasted bytes on the network; more complicated codegen) more than it helps (occasionally helps CPU performance; occasionally helps compression ratio). The IDL also supports generic types, which in Rust I'd like to codegen into Rust generic types.

I want to use `zerocopy` to eliminate all `unsafe` from the IDL-generated Rust code. This requires that `#[derive(AsBytes)]` works for generic types, and because of my choice to make all types unaligned, it should in principle be possible with this issue resolved.

FWIW, I originally started with `bytemuck` to hide my `unsafe` code. I ran into the same issue with lack of generic types, but unfortunately `bytemuck` doesn't offer the `Unaligned` trait and so they don't have such a simple path forward (https://github.com/Lokathor/bytemuck/issues/222 is the best I could come up with). Because of `Unaligned`, zerocopy seems much closer to what I'm looking for.

**Comment by joshlf** [comment]
Okay, sounds good.

I don't have time for this right now, but hopefully soon. I gave a bit of thought to what it'd take to implement, and it seems pretty straightforward. Leaving this breadcrumb for myself, I think we'd need to modify this block in `derive_as_bytes_struct`:

https://github.com/google/zerocopy/blob/431069fafef91879e65279f95d21aa26eb617b84/zerocopy-derive/src/lib.rs#L632-L638

In particular, we'd want to turn it into a match that can allow a number of combinations:
- Any struct with `repr(transparent)` or `repr(packed)` (this is already implemented)
- `repr(C)` so long as it only has one field
- Any type with no type parameters (in this case, we need to emit the padding check; this is already implemented)
- A type which doesn't satisfy any of the above (in this case, we need to emit `Unaligned` bounds on *all* fields both generic and concrete)

**Comment by joshlf** [comment]
@reinerp This has been released in 0.8.0-alpha.5. Thanks again for the suggestion!
